### PR TITLE
GEOMESA-1130 Users of KafkaDataStore should be able to specify a filt…

### PR DESCRIPTION
**Pull request by anthonyccri**
…er when consuming features

* Added a method to the data store to subscribe to a feature source with
  a filter.

Signed-off-by: Anthony Fox <anthonyfox@ccri.com>

### CODE REVIEW

File: geomesa-kafka/geomesa-kafka-datastore/src/main/scala/org/locationtech/geomesa/kafka/LiveKafkaConsumerFeatureSource.scala [hunk 1]
```diff
@@ -128,8 +128,10 @@ class LiveKafkaConsumerFeatureSource(entry: ContentEntry,
     while (running.get) {
       queue.take() match {
         case update: CreateOrUpdate =>
-          fireEvent(KafkaFeatureEvent.changed(this, update.feature))
-          featureCache.createOrUpdateFeature(update)
+          if(query.getFilter.evaluate(update.feature)) {
```
Comment (cne1x) [inline geomesa-kafka/geomesa-kafka-datastore/src/main/scala/org/locationtech/geomesa/kafka/LiveKafkaConsumerFeatureSource.scala]: Would this same filter check also apply to deletes?

File: geomesa-kafka/geomesa-kafka-datastore/src/main/scala/org/locationtech/geomesa/kafka/KafkaDataStore.scala [hunk 2]
```diff
@@ -39,15 +40,28 @@ class KafkaDataStore(override val zookeepers: String,
...
+  def getFeatureSource(typeName: String, filt: Filter): ContentFeatureSource = {
+    val entry = ensureEntry(new NameImpl(typeName))
+    val query = new Query(typeName, filt)
+    query.setStartIndex(0)
+    createFeatureSource(entry, query)
+  }
+
+  override def createFeatureSource(entry: ContentEntry) = createFeatureSource(entry, Query.ALL)
+  private def createFeatureSource(entry: ContentEntry, query: Query) = {
+    if(query.getStartIndex == null) query.setStartIndex(0)
```
Comment (elahrvivaz) [inline geomesa-kafka/geomesa-kafka-datastore/src/main/scala/org/locationtech/geomesa/kafka/KafkaDataStore.scala]: probably shouldn't use Query.ALL - query.getStartIndex will always be null, and then query.setStartIndex will throw an unsupported operation exception:
https://github.com/geotools/geotools/blob/master/modules/library/api/src/main/java/org/geotools/data/ALLQuery.java#L59
https://github.com/geotools/geotools/blob/master/modules/library/api/src/main/java/org/geotools/data/ALLQuery.java#L224

File: geomesa-kafka/geomesa-kafka-datastore/src/main/scala/org/locationtech/geomesa/kafka/LiveKafkaConsumerFeatureSource.scala (hunk 1 above)
Comment (anthonyccri) [inline geomesa-kafka/geomesa-kafka-datastore/src/main/scala/org/locationtech/geomesa/kafka/LiveKafkaConsumerFeatureSource.scala]: Deletes do not have the feature . . . only the feature id.

File: geomesa-kafka/geomesa-kafka-datastore/src/main/scala/org/locationtech/geomesa/kafka/KafkaDataStore.scala (hunk 2 above)
Comment (anthonyccri) [inline geomesa-kafka/geomesa-kafka-datastore/src/main/scala/org/locationtech/geomesa/kafka/KafkaDataStore.scala]: Ok, i'll change it to be a local 'All' and check for 'All'

File: geomesa-kafka/geomesa-kafka-datastore/src/main/scala/org/locationtech/geomesa/kafka/LiveKafkaConsumerFeatureSource.scala (hunk 1 above)
Comment (cne1x) [inline geomesa-kafka/geomesa-kafka-datastore/src/main/scala/org/locationtech/geomesa/kafka/LiveKafkaConsumerFeatureSource.scala]: That's right.  I forget that we only get the DELETE filter back.

From the end-users' point of view, then...
1.  an INSERT/UPDATE will send me an event iff it satisfies my filter;
2.  a DELETE will always sent me an event, even if what was removed does not satisfy my filter

Am I understanding this correctly?

**Comment by cne1x** [comment]
Building this now; if it succeeds, I'll merge this PR.

**Comment by cne1x** [comment]
Merge-squashed as commit b64db18b2fd9430e710874dff29252b38a135e0c

**Pull request closed without merging**